h2o-algos/src/test/java/hex/gam/GAMISplineTest.java

```diff
@@ -0,0 +1,346 @@
+            DKV.put(validGamified);
+            Scope.track(validGamified);
+            Frame trainGamified = DKV.getGet(gam._output._gamTransformedTrainCenter);
+            Scope.track(trainGamified);
+            TestUtil.assertIdenticalUpToRelTolerance(validGamified, trainGamified, EPS);
+        } finally {
+            Scope.exit();
+        }
+    }
+
+    /**
+     * This test will check the penalty matrix calculation for order = 2, 3, 4.  The manually derived penalty matrix
+     * will be centralized by multiplying with z-matrix as Z'*p*Z.  In addition, I will make sure of the following
+     * items that have been tested before to arrive at the manually derived matrix:
+     * 1. derivative of coefficients;
+     * 2. polynomial multiplications;
+     * 3. integration of polynomials;
+     * 4. derivation of z-matrix
+     */
+    @Test
+    public void testPenaltyMatrix() {
+        Scope.enter();
+        try {
+            Frame train = Scope.track(generateRealWithRangeOnly(4, 100, 0, 12345,
+                    1)); // generate training frame
+            // generate knots frame
+            double[][] pctiles = new double[][]{{-1}, {-0.5}, {0}, {0.5}, {1}};
+            Frame knotsFrame1 = genFrameKnots(pctiles);
+            DKV.put(knotsFrame1);
+            Scope.track(knotsFrame1);
+            // generate gamified frame
+            String[][] gamCols = new String[][]{{"C1"}, {"C2"}, {"C3"}};
+            GAMModel.GAMParameters params = new GAMModel.GAMParameters();
+            params._scale = new double[]{0.1, 0.1, 0.1};
+            params._bs = new int[]{2, 2, 2};
+            params._family = gaussian;
+            params._response_column = "C4";
+            params._spline_orders = new int[]{2, 3, 4};
+            params._max_iterations = 1;
+            params._savePenaltyMat = true;
+            params._gam_columns = gamCols;
+            params._knot_ids = new String[]{knotsFrame1._key.toString(), knotsFrame1._key.toString(), 
+                    knotsFrame1._key.toString()};
+            params._train = train._key;
+            params._solver = GLMModel.GLMParameters.Solver.IRLSM;
+            params._keep_gam_cols = true;
+            final GAMModel gam = new GAM(params).trainModel().get();
+            Scope.track_generic(gam);
+            // manually generating penalty matrix and check for order = 2
+            double[][] penaltyMat2 = new double[][]{{32, -16, 0, 0, 0}, {-16, 16, -8, 0, 0}, {0, -8, 16, -8, 0}, {0, 0,
+                    -8, 16, -16}, {0, 0, 0, -16, 32}}; // manually derived.
+            ArrayUtils.mult(penaltyMat2, 1.0 / gam._output._penaltyScale[0]);
+            assert2DArrayEqual(penaltyMat2, gam._output._penaltyMatrices[0]);
+            // manual penalty matrix check for order = 3
+            double[][] penaltyMat3 = new double[][]{{96, -36, -8, 0, 0, 0}, {-36, 24, -2, -2, 0, 0}, {-8, -2, 8, -2 - 2.0 / 3, -2, 0},
```

Comment:
In Gam_Monotone_Spline.pdf this penalty matrix for order 3 is wrong (eq 19 page 12). Would be worth fixing it there to avoid any future confusion in case of any problems.

Reply:
I thought it is just a few mistake but no, almost the whole thing is wrong.  I fixed it in the doc.  Thanks, Wendy